onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1982,22 +1982,25 @@ void RegisterContribSchemas() {
             AttributeProto::INT, static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT))
       .AllowUncheckedAttributes()
       .Input(0, "X", "Input data tensor from the previous layer.", "T")
-      .Input(1, "Scale", "Scale tensor.", "T")
-      .Input(2, "B", "Bias tensor.", "T", OpSchema::Optional)
-      .Output(0, "Y", "Output data tensor.", "T")
+      .Input(1, "Scale", "Scale tensor.", "V")
+      .Input(2, "B", "Bias tensor.", "V", OpSchema::Optional)
+      .Output(0, "Y", "Output data tensor.", "V")
```

Comment:
>V [](http://example.com/codeflow?start=46&length=1)

Inputs of type (T, V) will produce type (V and U). 
So for cases T=float, V=half.
We would allow Input <T=float, V=half> --> Output <V=half, U=float>. 

One of the problem is, the intermediate results ideally need a safe data type (after input type promotion) , do we have some logics  gurantee this in the kernel implementation?

Reply:
U is the type for calculation internal, normally float (double for double T and V). Previously we require input, weight and bias are all T, T can be float or half, now we can support T for input, V for weight and bias, they can be different, and are are smaller than U in size. The output type is same as weight and bias, which is V. During the calculation, we will cast data to U for calculation internally, so we will not loss precision. So for kernel registration, we don't support U as half. Actually this kernel is copied from apex, but from an older version which have T and U only. Apex already updated their code to support T, V and U, I just followed the same way to move the new version of code from Apex to here... The code is here: https://github.com/NVIDIA/apex/blob/master/csrc/layer_norm_cuda_kernel.cu

Reply:
I see, thanks for the explanation!